### ISSUE TITLE
Improve priv_msg/N

### ISSUE THREAD

**Issue by eazar001** [eazar001/yesbot#17]
I have  priv_msg/2 predicate that will convert messages with newlines into separate messages.
For example:

```
?- priv_msg("first\nsecond", "##prolog").
yesbot | first
yesbot | second
```

This is fine, however for certain cases where newlines _are not built in_, I would like to include a variant of priv_msg/N that would insert newlines for the user. This is because IRC has a 510 char limit that is initially reduced by the user's prefix. I should come up with some variants that optionally autoformat messages for the user. `priv_msg(Msg, Target, [auto_nl(true)])`, is one viable strategy, for example. Of course, keeping the original `priv_msg/2` would be a good idea.

**Comment by Anniepoo** [comment]
as a _user_ of priv_msg it's best if I don't have to know about limits of the IRC protocol.
most users expect priv_msg to post the message in the channel, without getting a bunch of stuff about IRC protocols involved in their thinking.
So, I'd say, a better contract for priv_msg/2 is 
1. if there are newlines in the message, convert them to separate raw messages.
2. if there are lines resulting from the above that are > the 510 less header info limit, cut them up. You can do this intelligently (look for word separators), or just compute the max line len, subtract a couple, and whack it off.

if you make a new pred whose only difference is cutting up long/nl messages longer how is the user to understand that contract? Aren't you making the contract complex?

**Comment by eazar001** [comment]
Right, and this is how I envisioned things, _sans_ the fully automated behavior you described. Reason I'm against this train of thought is:

spamming.

**Comment by eazar001** [comment]
Now understandably, there will be those desiring more control, and those desiring less. To make both camps happy, I intend to address this by creating a few different versions of `priv_msg`. In the end, those in tune with your philosophy will be happy, and those who are more paraonoid (like myself), will be put at ease too.

An example of this is the `wiki_search` extension which (thankfully) cuts off the message before it prints a half page html paragraph, to the channel. Admittedly, one would argue that the developer should handle this beforehand. Again, admittedly, i'm not 100% sure about this yet. So I'm still thinking about how to properly handle this, taking into consideration all persepectives and possibilities.

**Comment by eazar001** [comment]
One thought I just had about this was to allow the user an overall _message limit_ option in `priv_msg/N`. This might get around the issue of unpredictable spam.

I'm pretty much liking the idea of:
`priv_msg/3` as `priv_msg(+String, +Recipient, :ListOptions)`

and keeping `priv_msg/2` that is a default which calls `priv_msg(String, Recipient, [])` (empty list of options)

Perhaps this makes my intentions clearer?

Maybe better than passing an 'empty list of options' would be to pass a reasonable set of defaults that is in line with what you suggested above.

**Comment by Anniepoo** [comment]
ok, that makes sense - I think you've got it.

I _do_ think 'the developer' should handle spam prevention. But there's nothing wrong with them handling it by calling the appropriate, length limiting API.

since there's other potential versions, yes, the options is the best, and per the usual convention priv_msg/2 expands to priv_msg/3 with nul option list.

other options i can imagine (just as proof this is best)  -  allow/disallow messages that start with slash, or only /me or some other restricted set, (this should, for example, be set for chatty, so users don't trick chatty into saying commands) , auto enclose in thought bubble . o O (  this is a silly example ) , unicode allow/disallow, finer control over above line handling, etc.

**Comment by eazar001** [comment]
A compendium of thoughts for options in `priv_msg/3` thus far:
- [x] 1) "Auto-newline" formatting that inserts newlines for message separation when text limits are exceeded.
- [x] 2) Message rate limiting (applicable when auto newline is selected) to prevent too many messages from being sent at once. Perhaps some form of time limit can be implemented, or even a quantity limit option on top of all this would also be appealing.

~~3) Disallow messages that begin with '/'. (maybe pass a prefix exclusion list?)~~
- [x] 3) Change text encoding (i.e. octet, utf8, etc.)
- [ ] 4) CTCP style actions (Optional, for later)

**Comment by Anniepoo** [comment]
I like 'disallow prefix list'  - yes, the encoding is a good idea.
It'd be cool if this predicate took 'text' (that is, what happy with 'this is a line', "this is a line", or `this will be codes`

**Comment by eazar001** [comment]
I'm going to scratch point (3) off the list, as this doesn't really fit under the purview of a predicate that should be dealing with formatting/printing. Logic before the action of `priv_msg/N` should deal with point (3).

**Comment by eazar001** [comment]
Resolved, by way of commit ba5768378674d4e734f34342773e052900f33b6a.
